File: src/core/services/entry.service.ts
import type { DatabaseContext } from '../database/connection.js'
import { EntryVersionQuery } from '../database/queries/entry-version.query.js'
import { ScopedEntryQuery } from '../database/queries/scoped-entry.query.js'
import { EntryRepository } from '../database/repositories/entry.repository.js'
import { EntryStatusRepository } from '../database/repositories/entry-status.repository.js'
import { VersionRepository } from '../database/repositories/version.repository.js'
import type { ScopedEntry } from '../database/types.js'

export class EntryService {
  private entryRepo: EntryRepository
  private statusRepo: EntryStatusRepository
  private versionRepo: VersionRepository
  private scopedEntryQuery: ScopedEntryQuery
  private entryVersionQuery: EntryVersionQuery

  constructor(private ctx: DatabaseContext) {
    this.entryRepo = new EntryRepository(ctx)
    this.statusRepo = new EntryStatusRepository(ctx)
    this.versionRepo = new VersionRepository(ctx)
    this.scopedEntryQuery = new ScopedEntryQuery(ctx)
    this.entryVersionQuery = new EntryVersionQuery(ctx)
  }

  async getLatest(scopeId: number, key: string): Promise<ScopedEntry | undefined> {
    return this.scopedEntryQuery.getLatest(scopeId, key)
  }

  async getByVersion(scopeId: number, key: string, version: number): Promise<ScopedEntry | undefined> {
    return this.scopedEntryQuery.getByVersion(scopeId, key, version)
  }

  async getNextVersion(scopeId: number, key: string): Promise<number> {
    return this.entryVersionQuery.getNextVersion(scopeId, key)
  }

  async create(
    entry: Omit<ScopedEntry, 'id' | 'createdAt' | 'isArchived'> & { isArchived?: boolean },
  ): Promise<number> {
    return this.ctx.db.transaction(() => {
      // Check if entry exists
      const existing = this.entryRepo.findByScopeAndKey(entry.scopeId, entry.key)

      let entryId: number

      if (existing) {
        entryId = existing.id

        // Check if entry_status exists, create if not
        const status = this.statusRepo.findByEntryId(entryId)
        if (!status) {
          this.statusRepo.create(entryId, entry.version, entry.isArchived)
        }
      } else {
        // Create new entry
        entryId = this.entryRepo.create(entry.scopeId, entry.key)

        // Create entry status
        this.statusRepo.create(entryId, entry.version, entry.isArchived)
      }

      // Insert version
      const versionId = this.versionRepo.create(entryId, entry.version, entry.filePath, entry.hash, entry.description)

      // Update current version
      this.statusRepo.updateCurrentVersion(entryId, entry.version)

      return versionId
    })()
  }

  async list(scopeId: number, includeArchived = false, allVersions = false): Promise<ScopedEntry[]> {
    return this.scopedEntryQuery.list(scopeId, includeArchived, allVersions)
  }

  async deleteVersion(scopeId: number, key: string, version: number): Promise<boolean> {
    return this.ctx.db.transaction(() => {
      const entry = this.entryRepo.findByScopeAndKey(scopeId, key)
      if (!entry) return false

      const result = this.versionRepo.deleteByEntryAndVersion(entry.id, version)

      // Update current version if necessary
      const maxVersion = this.versionRepo.getMaxVersion(entry.id)

      if (maxVersion > 0) {
        this.statusRepo.updateCurrentVersion(entry.id, maxVersion)
      }

      return result
    })()
  }

  async deleteAll(scopeId: number, key: string): Promise<boolean> {
    return this.ctx.db.transaction(() => {
      const entry = this.entryRepo.findByScopeAndKey(scopeId, key)
      if (!entry) return false

      // Delete all versions
      this.versionRepo.deleteAllByEntry(entry.id)

      // Delete status
      this.statusRepo.delete(entry.id)

      // Delete entry
      this.entryRepo.delete(entry.id)

      return true
    })()
  }

  async archive(scopeId: number, key: string): Promise<boolean> {
    const entry = this.entryRepo.findByScopeAndKey(scopeId, key)
    if (!entry) return false

    // Check if already archived
    const status = this.statusRepo.findByEntryId(entry.id)
    if (status?.isArchived) return false

    return this.statusRepo.setArchived(entry.id, true)
  }

  async restore(scopeId: number, key: string): Promise<boolean> {
    const entry = this.entryRepo.findByScopeAndKey(scopeId, key)
    if (!entry) return false

    // Check if already active (not archived)
    const status = this.statusRepo.findByEntryId(entry.id)
    if (!status?.isArchived) return false

    return this.statusRepo.setArchived(entry.id, false)
  }

  getEntryByKey(scopeId: number, key: string) {
    return this.entryRepo.findByScopeAndKey(scopeId, key)
  }
}

File: internal/services/entry_service.go
package services

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/choplin/vault.md/internal/database"
	sqldb "github.com/choplin/vault.md/internal/database/sqlc"
)

// ErrNotFound is returned when a requested entry is not found.
var ErrNotFound = errors.New("entry not found")

// EntryService exposes high-level operations for scoped entries using sqlc-generated queries.
type EntryService struct {
	ctx *database.Context
}

// NewEntryService creates a new EntryService.
func NewEntryService(ctx *database.Context) *EntryService {
	return &EntryService{
		ctx: ctx,
	}
}

// GetLatest retrieves the latest version of an entry.
func (s *EntryService) GetLatest(ctx context.Context, scopeID int64, key string) (*database.ScopedEntryRecord, error) {
	q, err := s.queries()
	if err != nil {
		return nil, err
	}

	row, err := q.GetScopedEntryLatest(ctx, sqldb.GetScopedEntryLatestParams{
		ScopeID: scopeID,
		Key:     key,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}

	record := database.ScopedEntryRecordFromRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description)
	return &record, nil
}

// GetByVersion retrieves a specific version of an entry.
func (s *EntryService) GetByVersion(ctx context.Context, scopeID int64, key string, version int64) (*database.ScopedEntryRecord, error) {
	q, err := s.queries()
	if err != nil {
		return nil, err
	}

	row, err := q.GetScopedEntryByVersion(ctx, sqldb.GetScopedEntryByVersionParams{
		ScopeID: scopeID,
		Key:     key,
		Version: version,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}

	record := database.ScopedEntryRecordFromRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description)
	return &record, nil
}

// GetNextVersion returns the next version number for an entry.
func (s *EntryService) GetNextVersion(ctx context.Context, scopeID int64, key string) (int64, error) {
	q, err := s.queries()
	if err != nil {
		return 0, err
	}

	row, err := q.FindEntryByScopeAndKey(ctx, sqldb.FindEntryByScopeAndKeyParams{
		ScopeID: scopeID,
		Key:     key,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return 1, nil
		}
		return 0, err
	}

	maxVersion, err := q.MaxVersionForEntry(ctx, row.ID)
	if err != nil {
		return 0, err
	}
	return maxVersion + 1, nil
}

// Create persists a new entry version, provisioning the entry/status rows as needed.
func (s *EntryService) Create(ctx context.Context, entry database.ScopedEntryRecord) (versionID int64, err error) {
	err = s.withTx(ctx, func(txCtx context.Context, q *sqldb.Queries) error {
		row, err := q.FindEntryByScopeAndKey(txCtx, sqldb.FindEntryByScopeAndKeyParams{
			ScopeID: entry.ScopeID,
			Key:     entry.Key,
		})

		var entryID int64
		switch {
		case err == nil:
			entryID = row.ID
		case errors.Is(err, sql.ErrNoRows):
			res, err := q.InsertEntry(txCtx, sqldb.InsertEntryParams{
				ScopeID: entry.ScopeID,
				Key:     entry.Key,
			})
			if err != nil {
				return err
			}
			entryID, err = res.LastInsertId()
			if err != nil {
				return err
			}
			isArchived := sql.NullInt64{Int64: 0, Valid: true}
			if entry.IsArchived {
				isArchived.Int64 = 1
			}
			if err := q.InsertEntryStatus(txCtx, sqldb.InsertEntryStatusParams{
				EntryID:        entryID,
				IsArchived:     isArchived,
				CurrentVersion: sql.NullInt64{Int64: entry.Version, Valid: true},
			}); err != nil {
				return err
			}
		default:
			return err
		}

		if err == nil {
			_, err = q.FindEntryStatusByEntryID(txCtx, entryID)
			if errors.Is(err, sql.ErrNoRows) {
				isArchived := sql.NullInt64{Int64: 0, Valid: true}
				if entry.IsArchived {
					isArchived.Int64 = 1
				}
				if err := q.InsertEntryStatus(txCtx, sqldb.InsertEntryStatusParams{
					EntryID:        entryID,
					IsArchived:     isArchived,
					CurrentVersion: sql.NullInt64{Int64: entry.Version, Valid: true},
				}); err != nil {
					return err
				}
			} else if err != nil {
				return err
			}
		}

		var description sql.NullString
		if entry.Description != nil {
			description = sql.NullString{String: *entry.Description, Valid: true}
		}

		res, err := q.InsertVersion(txCtx, sqldb.InsertVersionParams{
			EntryID:     entryID,
			Version:     entry.Version,
			FilePath:    entry.FilePath,
			Hash:        entry.Hash,
			Description: description,
		})
		if err != nil {
			return err
		}
		if versionID, err = res.LastInsertId(); err != nil {
			return err
		}

		return q.UpdateEntryStatusCurrentVersion(txCtx, sqldb.UpdateEntryStatusCurrentVersionParams{
			CurrentVersion: sql.NullInt64{Int64: entry.Version, Valid: true},
			EntryID:        entryID,
		})
	})
	if err != nil {
		return 0, err
	}
	return versionID, nil
}

// List retrieves entries from the vault with specified filters.
func (s *EntryService) List(ctx context.Context, scopeID int64, includeArchived, allVersions bool) ([]database.ScopedEntryRecord, error) {
	q, err := s.queries()
	if err != nil {
		return nil, err
	}

	if allVersions {
		rows, err := q.ListScopedEntriesAllVersions(ctx, sqldb.ListScopedEntriesAllVersionsParams{
			ScopeID:         scopeID,
			IncludeArchived: includeArchived,
		})
		if err != nil {
			return nil, err
		}

		result := make([]database.ScopedEntryRecord, 0, len(rows))
		for _, row := range rows {
			result = append(result, database.ScopedEntryRecordFromRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description))
		}
		return result, nil
	}

	rows, err := q.ListScopedEntriesLatest(ctx, sqldb.ListScopedEntriesLatestParams{
		ScopeID:         scopeID,
		IncludeArchived: includeArchived,
	})
	if err != nil {
		return nil, err
	}

	result := make([]database.ScopedEntryRecord, 0, len(rows))
	for _, row := range rows {
		result = append(result, database.ScopedEntryRecordFromRow(row.EntryID, row.ScopeID, row.Key, row.EntryCreatedAt, row.IsArchived, row.Version, row.FilePath, row.Hash, row.Description))
	}
	return result, nil
}

// DeleteVersion deletes a specific version of an entry and returns true if deleted.
func (s *EntryService) DeleteVersion(ctx context.Context, scopeID int64, key string, version int64) (bool, error) {
	var deleted bool
	err := s.withTx(ctx, func(txCtx context.Context, q *sqldb.Queries) error {
		row, err := q.FindEntryByScopeAndKey(txCtx, sqldb.FindEntryByScopeAndKeyParams{
			ScopeID: scopeID,
			Key:     key,
		})
		if err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				deleted = false
				return nil
			}
			return err
		}

		affected, err := q.DeleteVersionByEntryAndVersion(txCtx, sqldb.DeleteVersionByEntryAndVersionParams{
			EntryID: row.ID,
			Version: version,
		})
		if err != nil {
			return err
		}

		maxVersion, err := q.MaxVersionForEntry(txCtx, row.ID)
		if err != nil {
			return err
		}
		if maxVersion > 0 {
			if err := q.UpdateEntryStatusCurrentVersion(txCtx, sqldb.UpdateEntryStatusCurrentVersionParams{
				CurrentVersion: sql.NullInt64{Int64: maxVersion, Valid: true},
				EntryID:        row.ID,
			}); err != nil {
				return err
			}
		}

		deleted = affected > 0
		return nil
	})
	if err != nil {
		return false, err
	}
	return deleted, nil
}

// DeleteAll deletes all versions of an entry and returns true if deleted.
func (s *EntryService) DeleteAll(ctx context.Context, scopeID int64, key string) (bool, error) {
	var deleted bool
	err := s.withTx(ctx, func(txCtx context.Context, q *sqldb.Queries) error {
		row, err := q.FindEntryByScopeAndKey(txCtx, sqldb.FindEntryByScopeAndKeyParams{
			ScopeID: scopeID,
			Key:     key,
		})
		if err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				deleted = false
				return nil
			}
			return err
		}

		if _, err := q.DeleteVersionsByEntry(txCtx, row.ID); err != nil {
			return err
		}
		if _, err := q.DeleteEntryStatus(txCtx, row.ID); err != nil {
			return err
		}
		affected, err := q.DeleteEntryByID(txCtx, row.ID)
		if err != nil {
			return err
		}

		deleted = affected > 0
		return nil
	})
	if err != nil {
		return false, err
	}
	return deleted, nil
}

// Archive marks an entry as archived and returns true if archived.
func (s *EntryService) Archive(ctx context.Context, scopeID int64, key string) (bool, error) {
	q, err := s.queries()
	if err != nil {
		return false, err
	}

	entryRow, err := q.FindEntryByScopeAndKey(ctx, sqldb.FindEntryByScopeAndKeyParams{
		ScopeID: scopeID,
		Key:     key,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return false, nil
		}
		return false, err
	}

	statusRow, err := q.FindEntryStatusByEntryID(ctx, entryRow.ID)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return false, nil
		}
		return false, err
	}
	status := database.EntryStatusRecordFromRow(statusRow)
	if status.IsArchived {
		return false, nil
	}

	affected, err := q.UpdateEntryStatusArchived(ctx, sqldb.UpdateEntryStatusArchivedParams{
		IsArchived: sql.NullInt64{Int64: 1, Valid: true},
		EntryID:    entryRow.ID,
	})
	if err != nil {
		return false, err
	}
	return affected > 0, nil
}

// Restore unarchives an entry and returns true if restored.
func (s *EntryService) Restore(ctx context.Context, scopeID int64, key string) (bool, error) {
	q, err := s.queries()
	if err != nil {
		return false, err
	}

	entryRow, err := q.FindEntryByScopeAndKey(ctx, sqldb.FindEntryByScopeAndKeyParams{
		ScopeID: scopeID,
		Key:     key,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return false, nil
		}
		return false, err
	}

	statusRow, err := q.FindEntryStatusByEntryID(ctx, entryRow.ID)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return false, nil
		}
		return false, err
	}
	status := database.EntryStatusRecordFromRow(statusRow)
	if !status.IsArchived {
		return false, nil
	}

	affected, err := q.UpdateEntryStatusArchived(ctx, sqldb.UpdateEntryStatusArchivedParams{
		IsArchived: sql.NullInt64{Int64: 0, Valid: true},
		EntryID:    entryRow.ID,
	})
	if err != nil {
		return false, err
	}
	return affected > 0, nil
}

// GetEntryByKey retrieves the entry record for a given key.
func (s *EntryService) GetEntryByKey(ctx context.Context, scopeID int64, key string) (*database.EntryRecord, error) {
	q, err := s.queries()
	if err != nil {
		return nil, err
	}
	row, err := q.FindEntryByScopeAndKey(ctx, sqldb.FindEntryByScopeAndKeyParams{
		ScopeID: scopeID,
		Key:     key,
	})
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}
	record := database.EntryRecordFromRow(row)
	return &record, nil
}

func (s *EntryService) withTx(ctx context.Context, fn func(context.Context, *sqldb.Queries) error) error {
	if s.ctx == nil || s.ctx.DB == nil {
		return fmt.Errorf("entry service: missing database context")
	}

	tx, err := s.ctx.DB.BeginTx(ctx, nil)
	if err != nil {
		return err
	}

	queries := sqldb.New(tx)

	if err := fn(ctx, queries); err != nil {
		_ = tx.Rollback()
		return err
	}

	if err := tx.Commit(); err != nil {
		_ = tx.Rollback()
		return err
	}

	return nil
}

func (s *EntryService) queries() (*sqldb.Queries, error) {
	if s.ctx == nil {
		return nil, fmt.Errorf("entry service: missing database context")
	}
	if s.ctx.Queries == nil {
		if s.ctx.DB == nil {
			return nil, fmt.Errorf("entry service: database handle not initialised")
		}
		s.ctx.Queries = sqldb.New(s.ctx.DB)
	}
	return s.ctx.Queries, nil
}
